cmd/juju/cloud/add.go

```diff
@@ -56,15 +58,24 @@ type addCloudCommand struct {
 
 	// CloudFile is the name of the cloud YAML file.
 	CloudFile string
+
+	// Ping contains the logic for pinging a cloud endpoint to know whether or
+	// not it really has a valid cloud of the same type as the provider.  By
+	// default it just calls the correct provider's Ping method.
+	Ping func(p environs.EnvironProvider, endpoint string) error
```

Comment:
Maybe we should consider folding this and the now ~CloudMetadataStore~ interface into a general ~AddCloudAPI~ interface?